lib/Dialect/FIRRTL/Transforms/GrandCentralSignalMappings.cpp

```diff
@@ -0,0 +1,255 @@
+  mapping.dir = anno.getMember<StringAttr>("dir").getValue() == "source"
+                    ? MappingDirection::ProbeRemote
+                    : MappingDirection::DriveRemote;
+  mapping.remoteTarget = anno.getMember<StringAttr>("peer");
+  mapping.localValue = value;
+  mapping.type = value.getType().cast<FIRRTLType>();
+
+  // Guess a name for the local value. This is only for readability's sake,
+  // giving the pass a hint for picking the names of the generated module ports.
+  if (auto blockArg = value.dyn_cast<BlockArgument>()) {
+    mapping.localName =
+        module.portNames()[blockArg.getArgNumber()].cast<StringAttr>();
+  } else if (auto op = value.getDefiningOp()) {
+    mapping.localName = op->getAttrOfType<StringAttr>("name");
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "  - " << mapping << "\n");
+  mappings.push_back(std::move(mapping));
+}
+
+void ModuleSignalMappings::emitMappingsModule() {
+  LLVM_DEBUG(llvm::dbgs() << "- Generating `" << mappingsModuleName << "`\n");
+
+  // Determine what ports this module will have, given the signal mappings we
+  // are supposed to emit.
+  SmallVector<ModulePortInfo> ports;
+  for (auto &mapping : mappings) {
+    ports.push_back(ModulePortInfo{mapping.localName, mapping.type,
+                                   mapping.dir == MappingDirection::DriveRemote
+                                       ? Direction::Input
+                                       : Direction::Output,
+                                   module.getLoc()});
+    LLVM_DEBUG(llvm::dbgs() << "  - Adding port " << mapping.localName << "\n");
+  }
+
+  // Create the actual module.
+  ImplicitLocOpBuilder builder(module.getLoc(), module);
+  auto mappingsModule = builder.create<FModuleOp>(
+      StringAttr::get(module.getContext(), mappingsModuleName), ports);
+
+  // Generate the connect and force statements inside the module.
+  builder.setInsertionPointToStart(mappingsModule.getBodyBlock());
+  unsigned portIdx = 0;
+  for (auto &mapping : mappings) {
+    // TODO: Actually generate a proper XMR here. For now just do some textual
+    // replacements. Generating a real IR node (like a proper XMR op) would be
+    // much better, but the modules that `EmitSignalMappings` interacts with
+    // generally live in a separate circuit. Multiple circuits are not fully
+    // supported at the moment.
+    auto circuitSplit = mapping.remoteTarget.getValue().split('|').second;
+    auto moduleSplit = circuitSplit.split('>');
+    SmallString<32> remoteXmrName(moduleSplit.first.split(':').first);
+    remoteXmrName.push_back('.');
+    for (auto c : moduleSplit.second) {
+      if (c == '[' || c == '.')
+        remoteXmrName.push_back('_');
+      else if (c != ']')
+        remoteXmrName.push_back(c);
+    }
+    if (mapping.dir == MappingDirection::DriveRemote) {
+      auto xmr = builder.create<VerbatimWireOp>(mapping.type, remoteXmrName);
+      builder.create<ForceOp>(xmr, mappingsModule.getArgument(portIdx));
```

Comment:
Do these `$force` statements need to be guarded with ifdefs? I've never run a GCT signal driving test, but the documentation seems to wrap these in `ifndefs` asserting that `VERILATOR` isn't defined?

Reply:
Yeah good catch, these are still missing.

Reply:
Done!